### ISSUE TITLE
Library has issues on esp8266 platform

### ISSUE THREAD

**Issue by bperrybap** [Nospampls/SchedTask#1]
Unfortunately, the esp8266 platform has a NOW() macro buried down deep in one of its internal headers that gets included when <Arduino.h> is included.
.arduino15/packages/esp8266/hardware/esp8266/2.7.1/tools/sdk/include/eagle_soc.h
This causes a compilation error for SchedBase.cpp
You can see this issue if you turn on warnings in the IDE and compile Example_1 for an esp8266 board like the NodeMCU 1.0 
This is a tough one to fix since it is macro conflict with a system header file and any change will break the API and documentation.
As a quick and dirty hack you can move the #include <Arduino.h> above the defines in SchedBase.h and then put a #undef NOW just above the define for NOW - to eliminate the redefinition warning.

A more general fix might be to create static const values for these defines and move them inside the ScheBase class.

i.e. inside the public section add:
```
    static const unsigned long NEVER 0xffffffff;
    static const unsigned long NOW = 0UL;
    static const unsigned long ONESHOT = 0UL;
```

Then to reference them in code you would use SchedBase::{symbolname}
i.e.
```
SchedTask OnTask (SchedBase::NOW, SchedBase::ONESHOT, turnOnLED);					// define the turn on task (dispatch now, one time)
SchedTask OffTask (SchedBase::NEVER, SchedBase::ONESHOT, turnOffLED);				// define the turn off task (dispatch
```

Not quite a simple as before, but there would never be a macro conflict issue with any other code.

**Comment by Nospampls** [comment]
Thanks for pointing this out as well as taking time to make suggestions.
I'm going to try to thread the needle here and not break the existing
documentation and API.  Please give me your assessment of these proposed
changes:

- I'll move #include <Arduino.h> up in SchedBase.h
- I'll precede the #define NOW 0UL with #undef NOW
- I'll create the following in the public: section of SchedBase.h and use
them internally

-- static const unsigned long SCHED_NOW = 0UL;
-- static const unsigned long SCHED_NEVER = 0xffffffff;
-- static const unsigned long SCHED_ONESHOT = 0UL;

- I'll update the documentation to identify these consts


I'm unsure of the impact of redefining NOW in this way on your platform.
Does it break?

I want to keep the API as simple as possible for the casual user but am
willing to try to help others as well.  I'm assuming that a knowledgeable
user who needed to could undefine my NOW, NEVER and ONESHOT and either use
his own #defines or the consts provided above.  Am I correct?

Thanks.

I'm assuming that a knowledgeable

**Comment by Nospampls** [comment]
I thought about this a little more.  How about I add new #defines for
SCHED_NOW, etc.?  Those would be safe and a user could undefine NOW, etc.

**Comment by bperrybap** [comment]
The issue as the code originally was is that #include <Arduino.h> was after the defines.
This would end up getting an error down in the core code. If a user didn't have warnings enabled in the IDE (which they aren't by default) all they will see is a compile error down in a system file and will give up.
I just saw this happen in an Arduino forum thread.
Keep in mind that the typical Arduino user often does not have much s/w skills so many will be unable to figure out what is happening much less be able to fix the issue with a work around.

In terms of a fix, there are several ways to go.
Each of the three proposed so far have advantages / disadvantages.
Which one you select is really up to you.
That said, whichever way you go, I would definitely move the #include<Arduino.h> up near the top so that any symbol collision issues will show up in the SchedBase code vs down in a platform system file.

1) move up #include<Arduino.h>  and add #undef NOW
This has the advantage that it is the simplest and no existing code or documentation has to change.
That definitely has value.
The disadvantage it is kind of awkard. But perhaps that is ok as it is all under the hood and users will not really see it.
I've had to do some similar ugly stuff like this in some of my libraries on various cores that have implementation issues related to symbols & macros for  LED_BUILTIN, SDA, SCL, INPUT_PULLUP
If the #undef is a concern, you could only to the #undef for NOW on the esp platforms.


2)  add consts to the SchedBase class
The advantage of this is that there will never be any sort of conflict with any other code or its macros.
The disadvantage is that it is not backward compatible with existing code and documentation.
Also the symbol references may look strange to some users.
A modification of this might be to do both 1 and 2 and update documentation and examples for using the SchedBase const vlues. This could provide a transition period to keep existing code workin but new code would use the class const value.

3) create new maco symbol names for NOW, NEVER, ONESHOT
This has the advantage that it keeps the names a bit simpler.
The disadvantage is that it breaks all existing code and documentation.

**Comment by Nospampls** [comment]
Thanks.  I could use a bit more help if you're willing.  I'm ruling out 2
and 3 for the disadvantages you stated.  I'm confused about 1 -- if I undef
NOW won't that preclude my users from using NOW?  i.e. break the
documentation?

I thought about using a 'switch' in the main sketch to be used only for esp
platforms.  I'm not sure how to detect the platform at pre-compile time, so
I tried this:

in the .INO sketch:

#define USING_ESP8266


in SchedBase.h:

#include <Arduino.h>

// sched task value for next = now (dispatch immediately)
*#ifndef USING_ESP8266*
*#define NOW 0UL*
*#endif*
#define SCHED_NOW 0UL

// sched task value for next = never (wait for a change to dispatch)
#define NEVER 0xFFFFFFFF
#define SCHED_NEVER 0xffffffff


ESP users should use the switch and SCHED_NOW, etc. as does the modified
SchedBase.cpp.

HOWEVER, I can't make this work.  I can't prevent the #define NOW from
defining and I still get the redefinition error.  Either I'm missing
something that should be obvious or the pre-processor directives aren't
working properly (doubtful).

Can you help?

Thanks.

**Comment by bperrybap** [comment]
> I'm confused about 1 -- if I undef NOW won't that preclude my users from using NOW? i.e. break the documentation? 

No. You undef it before you define it to remove the previous definition.

Here is a code fragment from SchedBase.h to to show how that works:

```
#ifndef SchedBase_h
#define SchedBase_h

#include <Arduino.h>

// sched task value for next = never (wait for a change to dispatch)
#define NEVER 0xFFFFFFFF

// The macro NOW exists on some platforms and gets defined by including <Arduino.h>
// since the platform NOW isn't used by this library, we simply undefine it
#if defined(ARDUINO_ARCH_ESP8266) || defined(ARDUINO_ARCH_ESP32) 
#undef NOW
#endif
// sched task value for next = now (dispatch immediately)
#define NOW 0UL

// sched task value for period = one shot (dispatch only once at t=next)
#define ONESHOT 0UL
```

The conditional for the undef for NOW is optional.
It can be used if you only want to do the undef on specific platorms.

ARDUINO_ARCH_ESP8266 and ARDUINO_ARCH_ESP32 are defined by the IDE when boards from those platforms are used.
using these builtin defines allows the library code to auto detect building the code for these platforms.

Note: I don't know if  NOW is an issue for the esp32 platform.
I tried to do a quick test on the esp32 platform, but there are several otherl issues that need to be resolved before the library can compile on the esp32 platform.
Fixing one of the issues was a little more than I wanted to tackle before getting back to you.

**Comment by Nospampls** [comment]
I tested this by compiling for the ESP and UNO with no errors (with the
conditional).  I think I finally get why this works -- the pre-compiler has
already used NOW() and no longer needs it for the platform code, right?

I'll also change addTask to *return ++taskCount;* though it's not used by
the user.  Maybe you caught that with compiler warnings enabled.

It sounds like you might find other SchedTask issues with the esp32 so I'll
wait for more from you before I release another version.

Finally, what do you think about the utility of my library compared to my
'competition'?  I'd value your comments since it's clear to me you have
more experience than I with Arduino and C++.  In my case I use it in
virtually every sketch I write.

Thanks.  I appreciate your patience and willingness to teach.

**Comment by bperrybap** [comment]
> I tested this by compiling for the ESP and UNO with no errors (with the conditional). I think I finally get why this works -- the pre-compiler has already used NOW() and no longer needs it for the platform code, right?

Not really.

The macro NOW is defined as function type macro for use by the lower level code in the esp8266 platform in the header file
.arduino15/packages/esp8266/hardware/esp8266/2.7.1/tools/sdk/include/eagle_soc.h

`#define NOW()                              TIMER_REG_READ(FRC2_COUNT_ADDRESS)`

There is some low level core code that uses
`NOW()`
It is timer related code.
sys_now() in ./tools/sdk/lwip/include/lwip/sys.h uses it, as well as some code in ./tools/sdk/lwip/src/core/timers.c
and then some of the networking core code calls sys_now() 

Macros are substituted at  preprocessor time just before compilation. The esp8266 version of NOW() must exist when their low level core code is compiled but does not need to exist when arduino sketches and arduino libraries are called.

Arduino sketches and libraries should not need  NOW() so  the SchedBase library can safely undefine the macro and redefine it. It is a but ugly but should work.

IMO, the esp8266 core didn't need to include all the headers being included by Arduino.h or should have split the headers up a bit to keep some of the low level header files away from Arduino sketches and libraries.
i.e. they were a bit lazy with their headers
Also, they were a bit careless with their macro names.
IMO, they should not be using "normal" names like NOW.
Typically for internal under the hood macro names which is a global namespace, an underscore is inserted to avoid naming conflicts with user level code.
i.e. if they wanted a name like NOW they should have called it _NOW() or even __NOW() 

There are compilation issues with esp32 platform. I'll start another issue for that.

**Comment by Nospampls** [comment]
Thank you very much for the detailed explanation.   I'm eager to learn.
I'll reply to your other email re ESP32 separately after I review it in
detail.

--rolfe.

**Comment by Nospampls** [comment]
Please give version 1.0.0 a try.  It compiles for me though I cannot test it.

**Comment by Nospampls** [comment]
Please give version 1.1.0 a try.  Example_1 compiles for me.
